src/PointwiseFunctions/AnalyticSolutions/Poisson/ProductOfSinusoids.hpp

```diff
@@ -0,0 +1,85 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <limits>
+
+#include "DataStructures/DataBox/Prefixes.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/Tensor.hpp"     // IWYU pragma: keep
+#include "Elliptic/Systems/Poisson/Tags.hpp"    // IWYU pragma: keep
+#include "Options/Options.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// \cond
+class DataVector;
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace Poisson {
+namespace Solutions {
+/*!
+ * \brief A product of sinusoids \f$u(\boldsymbol{x}) = \prod_i \sin(k_i x_i)\f$
+ *
+ * \details Solves the Poisson equation \f$-\Delta u(x)=f(x)\f$ for a source
+ * \f$f(x)=\boldsymbol{k}^2\prod_i \sin(k_i x_i)\f$.
+ */
+template <size_t Dim>
+class ProductOfSinusoids {
+ public:
+  struct WaveNumbers {
+    using type = std::array<double, Dim>;
+    static constexpr OptionString help{"The wave numbers of the sinusoids"};
+  };
+
+  using options = tmpl::list<WaveNumbers>;
+  static constexpr OptionString help{"A product of sinusoids"};
+
+  ProductOfSinusoids() = default;
+  ProductOfSinusoids(const ProductOfSinusoids&) noexcept = delete;
+  ProductOfSinusoids& operator=(const ProductOfSinusoids&) noexcept = delete;
+  ProductOfSinusoids(ProductOfSinusoids&&) noexcept = default;
+  ProductOfSinusoids& operator=(ProductOfSinusoids&&) noexcept = default;
+  ~ProductOfSinusoids() noexcept = default;
+
+  explicit ProductOfSinusoids(
+      const std::array<double, Dim>& wave_numbers) noexcept;
+
+  tuples::TaggedTuple<Field, AuxiliaryField<Dim>> field_variables(
+      const tnsr::I<DataVector, Dim>& x) const noexcept;
+
+  tuples::TaggedTuple<::Tags::Source<Field>,
+                      ::Tags::Source<AuxiliaryField<Dim>>>
+  source_variables(const tnsr::I<DataVector, Dim>& x) const noexcept;
+
+  // clang-tidy: no pass by reference
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+  constexpr std::array<double, Dim> wave_numbers() const noexcept {
```

Comment:
why not return a reference to the array? Considering the constructors aren't constexpr, I don't see how this function could ever be evaluated in a constexpr context. When considering it has options to be read from an input file, it seems even less likely :)

Reply:
Fixed it. But why are e.g. the member accessors of `IsentropicVortex.hpp` `constexpr`?

Reply:
I'm not sure why those are. It's not wrong to have it be constexpr, but it would make sense to in that case have a constexpr constructor, etc. That is, "constexpr ALL the things" :)